### ISSUE TITLE
Add shuffle of data and test with non-default solvers

### ISSUE THREAD

**Issue by snath-xoc** [snath-xoc/sample-weight-audit-nondet#25]
I was initially a bit confused as to why all the deterministic estimators were passing all of a sudden. Looking into LogisticRegression, I realised it was failing due to the data being shuffled (as specified by the random_state description i.e., when the solver is one of ```liblinear, sag or saga```). It may be worth:

1.  Adding a feature to shuffle the data when no random state is found within the signature of est.__init__ (although in this case we are probably not testing for sample weight equivalence any more but rather just convergence?)
2. Testing the deterministic estimators with a range of solvers including at least lbfgs, liblinear and saga.

@ogrisel @antoinebaker @jeremiedbb let me know your thoughts

**Comment by snath-xoc** [comment]
O.K. trying (1) out doesn't change anything (phew), probably at some point implementing (2) would be good given the LogisticRegression test failures (however I feel like it is again looking more at a convergence issue if anything)

**Comment by antoinebaker** [comment]
Maybe I misunderstood your point, but I don't think the issue is data shuffling. 

For me `LogisticRegression(solver="sag")` and `LogisticRegression(solver="saga")` are stochastic estimators which subsample the dataset and should be tested with the statistical equivalence test only, they are not expected to pass the deterministic test.  `LogisticRegression(solver="lbfgs")`is a deterministic estimator which should pass the deterministic test. Is it what you have currently in the test suite ?

I'm not familiar with the C++ `liblinear` solver and where the randomization is used.

**Comment by antoinebaker** [comment]
Ah thanks for the clarification @snath-xoc , so I guess the real issue is that we have a genuine bug for sample_weight in liblinear, sag and saga solvers ?

**Comment by snath-xoc** [comment]
@antoinebaker after further investigation I created the following minimal reproducer to isolate the effects of sag:

```python
import numpy as np
from scipy.stats import kstest
from sklearn.linear_model.tests.test_sag import sag, squared_dloss
from sklearn.datasets import make_regression
from sklearn.utils._testing import assert_allclose_dense_sparse

step_size=1
alpha=0.01
n_features = 10

rng = np.random.RandomState(0)
X, y = make_regression(n_samples=1000,random_state=0,n_features=n_features)
weights = rng.randint(0,5,size=X.shape[0])

X_repeated = np.repeat(X,weights,axis=0)
y_repeated = np.repeat(y,weights,axis=0)

weights_w_all = np.zeros([n_features,100])
weights_r_all = np.zeros([n_features,100])

for random_state in np.arange(100):

    weights_w, int_w = sag(X,y,step_size=step_size,alpha=alpha,sample_weight=weights,dloss=squared_dloss)
    weights_w_all[:,random_state] = weights_w
    weights_r, int_r = sag(X_repeated,y_repeated,step_size=step_size,alpha=alpha,dloss=squared_dloss)
    weights_r_all[:,random_state] = weights_r

print(kstest(weights_r_all[0],weights_w_all[0]))

```

which gives the following output

```
KstestResult(statistic=np.float64(1.0), pvalue=np.float64(2.2087606931995054e-59), statistic_location=np.float64(2072040.1927989526), statistic_sign=np.int8(1))
```

Upon closer inspection of the sag (in the _saf_fast.pyx file, I see that the way sample weight is factored in is by multiplying the corrected gradient with the sample weight. Perhaps it would make more sense to factor in the sample weights when sampling indices over which to calculate the gradient i.e.,  ```python sample_ind ``` in L230

**Comment by antoinebaker** [comment]
@snath-xoc for the failures of "sag(a)" solvers, did you check that the fit has converged ? 
We expect that the repeated/weighted dataset will converge (after enough iterations) to the same solution (which is the actual minima) but in general we cannot expect the optimization paths to be the same, quoting  https://github.com/scikit-learn/scikit-learn/pull/30143#discussion_r1910470902

>the optimization paths for the weighted vs reweighted problems might not be equal for stochastic gradient (or minibatch) solvers.

>This would be the case for LogisticRegression / Ridge and the likes with the "sag" and "saga" solvers. It would also be the case for SGDClassifier/SGDRegressor with convex objective functions (I think all the choices of loss and penalties we implement lead to convex objectives, although convergence via SGD might not be guaranteed for non-smooth penalties).

A possible issue is that the weighted version in "saga(a)" may fail to converge, eg https://github.com/scikit-learn/scikit-learn/issues/21305. But I would say they are two distinct issues:
1. make sure the "sag(a)" solver, when it converges, converges towards to the true minima (which is the same for the weighted/repeated datasets with the proper rescaling of `alpha` or `C`). That means we must adapt our statistical test to make sure the weighted/repeated estimators reached convergence. 
2. fix convergence issues introduced by `sample_weight`

**Comment by ogrisel** [comment]
I did a quick test with the inspection notebook and indeed, it seems that we have convergence problems for large values of `C` (e.g. 1e3, low regularization) even with `max_iter=1_000_000`.

For small values of `C` (e.g. 1e-3 high regularization), there is no convergence problem and the statistical test passes (large p-value).

What remains to be investigated is whether the convergence problems happen for the weighted or repeated fit (or both). The fact that this solver does not converge on this dataset should probably be reported as a bug in itself, ideally with a minimal reproducer.

**Comment by antoinebaker** [comment]
> 2\. sample weights makes convergence more difficult overall

I think it might be the issue. According to https://en.wikipedia.org/wiki/Stochastic_variance_reduction a key ingredient for convergence in "sag" is that the $f_i$ have comparable condition number / Lipschitz smoothness, which I think is not the case for the weighted version where $f_i = w_i * loss_i$ is the loss at sample i multiplied by its sample weight.

**Comment by antoinebaker** [comment]
> What remains to be investigated is whether the convergence problems happen for the weighted or repeated fit (or both)

In my experiments "sag(a)" only struggle with the weighted dataset.

**Comment by ogrisel** [comment]
One possible fix would be to implement a minibatch version of SAG/SAGA and sample the minibatch contents using the weights with replacement (when provided). However, this would be significant refactoring work. Not sure if it's worth investing in this.


I wonder if @josephsalmon and co-authors have considered the weighted case when working on https://arxiv.org/abs/1902.00071.

**Comment by snath-xoc** [comment]
Shall we switch the conversation to PR [#31536](https://github.com/scikit-learn/scikit-learn/issues/31536)? I added the test and some more documentation there.

**Comment by josephsalmon** [comment]
> One possible fix would be to implement a minibatch version of SAG/SAGA and sample the minibatch contents using the weights with replacement (when provided). However, this would be significant refactoring work. Not sure if it's worth investing in this.
> 
> I wonder if [@josephsalmon](https://github.com/josephsalmon) and co-authors have considered the weighted case when working on https://arxiv.org/abs/1902.00071.

I don't think so, but @ngazagna is the right person for answering that.
